core/src/main/java/org/elasticsearch/common/logging/PrefixLogger.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.MarkerManager;
+import org.apache.logging.log4j.message.Message;
+import org.apache.logging.log4j.spi.ExtendedLogger;
+import org.apache.logging.log4j.spi.ExtendedLoggerWrapper;
+
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
+class PrefixLogger extends ExtendedLoggerWrapper {
+
+    private final String prefix;
+    private final Marker marker;
+
+    public String prefix() {
+        return prefix;
+    }
+
+    private static final WeakHashMap<String, WeakReference<Marker>> markers = new WeakHashMap<>();
+
+    PrefixLogger(final ExtendedLogger logger, final String name, final String prefix) {
+        super(logger, name, null);
+        this.prefix = prefix;
+
+        final String actualPrefix = (prefix == null ? "[] " : prefix).intern();
+        final Marker actualMarker;
+        synchronized (markers) {
+            final WeakReference<Marker> marker = markers.get(actualPrefix);
```

Comment:
>  (think of index and shard prefixes).

A sad, but valid point. add a comment?

> The synchronization is not related to the weak references, it's just to protect the map from concurrent modifications (which are not thread safe).

What happens if `marker.get()` returns non null at first call and a null at the second? should we capture it and make a hard reference to it?

Also to clarify - this method can not guarantee markers will not be created twice with the same prefix, but it limits it considerably, correct? 

Last - I'm still not convinced we shouldn't just go with creating a marker on each request and be done with it (it's on more object per logger, which I don't think is a huge deal).